res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+        }
+        output(value) {
+            const outval = this.outValueScale(value);
+            // NOTE: outputColor only handles hotcueColors
+            // and there is no hotcueColor for turning the LED
+            // off. So the `send()` function is responsible for turning the
+            // actual LED off.
+            if (this.colorKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorKey));
+            } else {
+                this.send(outval);
+            }
+        }
+        outputColor(colorCode) {
+            // Sends the color from the colorCode to the controller. This
+            // method will not be called if no colorKey has been specified.
+            if (colorCode === undefined || colorCode < 0 || colorCode > 0xFFFFFF) {
+                console.warn("Ignoring invalid color code '" + colorCode + "' in outputColor()");
+                return;
+            }
+
+            if (this.colorMapper !== undefined) {
+                // This HotcueButton holds a reference to a ColorMapper. This means
+                // that the controller only supports a fixed set of colors, so we
+                // get the MIDI value for the nearest supported color and send it.
+                const nearestColorValue = this.colorMapper.getValueForNearestColor(colorCode);
+                this.send(nearestColorValue);
+            } else {
+                // Since outputColor has been called but no ColorMapper is
+                // available, we can assume that controller supports arbitrary
+                // RGB color output.
+                this.sendRGB(colorCodeToObject(colorCode));
+            }
+        }
+        sendRGB(_colorObject) {
+            // This method needs to be overridden in controller mappings,
+            // because the procedure is controller-dependent.
+            throw "sendRGB(colorObject) not implemented -- unable to send RGB colors";
+        }
+        connect() {
+            super.connect();
+            if (this.group !== undefined && this.colorKey !== undefined) {
+                // TODO (Swiftb0y): replace with arrow function once https://bugreports.qt.io/browse/QTBUG-95677 got fixed
+                this.connections[1] = engine.makeConnection(this.group, this.colorKey, function(color) {
+                    if (engine.getValue(this.group, this.outKey)) {
+                        this.outputColor(color);
+                    }
+                }.bind(this));
+            }
+        }
+    };
+
+    const SamplerButton = class extends Button {
+        constructor(options) {
+            if (options.number === undefined) {
+                throw "No sampler number specified for SamplerButton constructor";
+            }
+            options.group = "[Sampler" + options.number + "]";
+            options.outKey = null; // hack to get Component constructor to call connect()
+            super(options);
```

Comment:
super constructors are always meant to be called before the child class does its own setup. And you've rightly identified that this is a hack.  Instead, can we think of a way to get this functionality to work without the hack?  (why can't we call connect() ourselves? I don't understand the code well enough)

Reply:
Neither do I. I spend a couple minutes looking into this and just calling `connect` does not do the trick. I think the comment is a bit misleading because it doesn't actually depend on `connect` being called in the constructor but some of the other stuff thats called between the check for `outKey === undefined` and `connect()`.

Reply:
thanks for taking a look at it.  We can leave this for now.